api/src/controllers/user.js

```diff
@@ -85,124 +86,92 @@ exports.list = (req, res) => {
+		return;
 	}
 
-	// TODO: go back in and clean this up (@kenhoff)
-	// first, check to see if the user exists.
-	User.findById(req.params.userId)
-		.then(user => {
-			if (!user) {
-				return res.sendStatus(404);
-			} else {
-				// (this gets a little wonky, just because we have to verify the user exists before trying to set all the follow relationship stuff, and we also want to only do the following if they pass in interests as part of the PUT request)
-				// if you know of a better way to make this one work, feel free to refactor!
-				return new Promise(resolve => {
-					if (data.interests) {
-						// then, if interests are provided as part of the PUT request - if so, go through and follow all the rss feeds / podcasts listed under each one of the interests
-						// (in the future, can do a quick diff to see if some of the interests are previous interests, or if the rss feeds / podcasts have already been followed - but for now, just refollow all the "interest" rss feeds / podcasts if the "interests" key is provided)
-
-						// for each "interest" provided:
-						return Promise.all([
-							data.interests.map(interest => {
-								// find all rss feeds and podcasts for that interest, and follow them
-								return Promise.all([
-									RSS.find({ interest }).then(rssFeeds => {
-										return Promise.all(
-											rssFeeds.map(rssFeed => {
-												return followRssFeed(req.params.userId, rssFeed._id);
-											}),
-										);
-									}),
-									Podcast.find({ interest }).then(podcasts => {
-										return Promise.all(
-											podcasts.map(podcast => {
-												return followPodcast(req.params.userId, podcast._id);
-											}),
-										);
-									}),
-								]);
-							}),
-						]).then(() => {
-							resolve();
-						});
-					} else {
-						resolve();
-					}
-				}).then(() => {
-					// update the user
-					User.findByIdAndUpdate({ _id: req.params.userId }, data, {
-						new: true,
-					}).then(user => {
-						// send back the user
-						user.password = undefined;
-						user.recoveryCode = undefined;
-
-						return res.json(user);
-					});
-				});
-			}
-		})
-		.catch(err => {
-			logger.error(err);
-			res.status(500).send(err);
+	let user = await User.findById(req.params.userId);
+
+	if (!user) {
+		res.sendStatus(404);
+		return;
+	}
+
+	if (data.username) {
+		// check for existing username
+		let userByUsername = await User.findOne({username: data.username});
+		if (userByUsername && userByUsername.id != user.id) {
+			res.status(409).send('User with this username already exists');
+			return;
+		}
+	}
+	if (data.email) {
+		// check for existing email
+		let userByEmail = await User.findOne({email: data.email});
+		if (userByEmail && userByEmail.email != user.email) {
+			res.status(409).send('User with this email already exists');
+			return;
+		}
+	}
+
+	if (data.interests) {
+		await data.interests.map(interest => {
```

Comment:
There's nothing to `await` resulting from the execution of this `map` as you are not returning anything from the function passed to map.
Its better to use a `for` loop here and `await` for results from `RSS` and `Podcast` collections in the loop body.